### PR TITLE
Dev | Program.cs Refactor

### PR DESCRIPTION
**Pull request by bseptember**
### Overview:

The Main method was over 200 lines. We needed to refactor the setup code.
Place them in their own related methods so its not hard to follow the code.

**Note: Please test on your side before approval.**

[Dev 9572](
https://dev.azure.com/retro-rabbit/RetroGradOnboard/_boards/board/t/RetroGradOnboard%20Team/Stories?System.AssignedTo=%40me&workitem=9572)

### CODE REVIEW

**Review by Wimmas — APPROVED** [review]

File: RR.App/Program.cs [hunk 1]
```diff
@@ -67,168 +76,115 @@ public static async Task Main(params string[] args)
...
-                builder.Services.AddAuthorization(options =>
+                services.AddAuthorization(options =>
                 {
-                    options.AddPolicy(
-                        policySettings.Name,
-                        policy =>
+                    options.AddPolicy(policySettings.Name, policy =>
+                    {
+                        policy.RequireRole(policySettings.Roles);
+                        if (policySettings.Permissions?.Any() == true)
```
Comment (EduanO) [inline RR.App/Program.cs:140]:
```suggestion
                        if (policySettings.Permissions.Any() == true)
```

File: RR.App/Program.cs [hunk 2]
```diff
@@ -67,168 +76,115 @@ public static async Task Main(params string[] args)
...
-            {
-                Policies = policies
-                .Select(policy => new PolicySettings {
-                    Name = policy.Value["Name"].First(),
-                    Roles = policy.Value["Roles"],
-                    Permissions = policy.Value["Permissions"]})
-
-                .ToList()
-            }.Policies.ForEach(policySettings =>
+            policies.ForEach(policySettings =>
```
Comment (EduanO) [inline RR.App/Program.cs:133]:
```suggestion
            policies?.ForEach(policySettings =>
```

File: RR.App/Program.cs [hunk 3]
```diff
@@ -67,168 +76,115 @@ public static async Task Main(params string[] args)
...
-                            if (jwksSet != null)
-                            {
-                                return jwksSet.Keys;
-                            }
-                            else
-                            {
-                                throw new InvalidOperationException("JsonWebKeySet is not available.");
-                            }
-                        }
+                            LazyJsonWebKeySet.Value?.Keys ?? throw new InvalidOperationException("JsonWebKeySet is not available.")
```
Comment (EduanO) [inline RR.App/Program.cs:109]:
```suggestion
                            LazyJsonWebKeySet.Value.Keys ?? throw new InvalidOperationException("JsonWebKeySet is not available.")
```

File: RR.App/Program.cs [hunk 4]
```diff
@@ -67,168 +76,115 @@ public static async Task Main(params string[] args)
...
                         ValidateAudience = true,
                         ValidateLifetime = true,
                         ValidateIssuerSigningKey = true,
                         ClockSkew = TimeSpan.Zero,
-
-                        ValidIssuer = Environment.GetEnvironmentVariable("AuthManagement__Issuer"),
-                        ValidAudience = Environment.GetEnvironmentVariable("AuthManagement__Audience"),
+                        ValidIssuer = authSettings["Issuer"],
+                        ValidAudience = authSettings["Audience"],
                         IssuerSigningKeyResolver = (token, securityToken, kid, validationParameters) =>
```
Comment (EduanO) [inline RR.App/Program.cs:108]:
```suggestion
                        IssuerSigningKeyResolver = (_, _, _, _) =>
```

File: RR.App/Program.cs (hunk 1 above)
Comment (EduanO) [inline RR.App/Program.cs:140]:
```suggestion
                         if (policySettings.Permissions.Any())
```

File: RR.App/Program.cs [hunk 5]
```diff
@@ -11,51 +11,60 @@
...
 
-        public static async Task Main(params string[] args)
+        public static async Task Main(string[] args)
         {
             var builder = WebApplication.CreateBuilder(args);
-            ConfigurationManager configuration = builder.Configuration;
-            configuration.AddJsonFile("appsettings.json");
+            var configuration = builder.Configuration;
+
+            ConfigureConfigurationSources(configuration);
```
Comment (EduanO) [inline RR.App/Program.cs:27]:
```suggestion
            SetupConfiguration(configuration);
```

File: RR.App/Program.cs [hunk 6]
```diff
@@ -11,51 +11,60 @@
...
+            ConfigureAuthentication(builder.Services, configuration);
+            ConfigureAuthorizationPolicies(builder.Services, configuration);
+
+            var app = builder.Build();
+            ConfigureApp(app);
+
+            await app.RunAsync();
+        }
+
+        private static void ConfigureConfigurationSources(ConfigurationManager configuration)
```
Comment (EduanO) [inline RR.App/Program.cs:39]:
```suggestion
        private static void SetupConfiguration(ConfigurationManager configuration)
```

File: RR.App/Program.cs [hunk 7]
```diff
@@ -11,51 +11,60 @@
...
+        {
+            configuration.AddJsonFile("appsettings.json", optional: false, reloadOnChange: true);
             configuration.AddUserSecrets<Program>();
+            configuration.AddEnvironmentVariables();
+        }
 
-            var serviceBusConnectionString = Environment.GetEnvironmentVariable("NewEmployeeQueue__ConnectionString");
-            var queueName = Environment.GetEnvironmentVariable("ServiceBus__QueueName");
+        private static void RegisterServices(IServiceCollection services, IConfiguration configuration)
```
Comment (EduanO) [inline RR.App/Program.cs:46]:
```suggestion
        private static void SetupDependencyInjection(IServiceCollection services, IConfiguration configuration)
```

File: RR.App/Program.cs [hunk 8]
```diff
@@ -11,51 +11,60 @@
...
-        public static async Task Main(params string[] args)
+        public static async Task Main(string[] args)
         {
             var builder = WebApplication.CreateBuilder(args);
-            ConfigurationManager configuration = builder.Configuration;
-            configuration.AddJsonFile("appsettings.json");
+            var configuration = builder.Configuration;
+
+            ConfigureConfigurationSources(configuration);
+            RegisterServices(builder.Services, configuration);
```
Comment (EduanO) [inline RR.App/Program.cs:28]:
```suggestion
            SetupDependencyInjection(builder.Services, configuration);
```

**Review by EduanO — CHANGES_REQUESTED** [review]
Please look at comments

**Review by MathildaShongwe15 — APPROVED** [review]

**Review by madwizzard — APPROVED** [review]
